server/context.go

```diff
@@ -170,13 +174,25 @@ func (s *SessionManager) NewContextWithQuery(conn *mysql.Conn, query string) (*s
 		sql.WithMemoryManager(s.memory),
 		sql.WithProcessList(s.processlist),
 		sql.WithRootSpan(s.tracer.StartSpan("query")),
+		sql.WithServices(sql.Services{
+			KillConnection: s.killConnection,
+		}),
 	)
 
 	return context, nil
 }
 
-// CloseConn closes the connection in the session manager and all its
-// associated contexts, which are cancelled.
+func (s *SessionManager) killConnection(connID uint32) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if entry, ok := s.sessions[connID]; ok {
+		delete(s.sessions, connID)
+		entry.conn.Close()
```

Comment:
Maybe add a note that we always kill any associated process before calling this (it looks incomplete as is)